src/ray/raylet/node_manager.cc

```diff
@@ -240,6 +255,36 @@ void NodeManager::ClientAdded(const ClientTableDataT &client_data) {
   remote_server_connections_.emplace(client_id, std::move(server_conn));
 }
 
+void NodeManager::ClientRemoved(const ClientTableDataT &client_data) {
+  const ClientID client_id = ClientID::from_binary(client_data.client_id);
+  RAY_LOG(DEBUG) << "[ClientRemoved] received callback from client id " << client_id;
+
+  // If the client has already been removed, don't do anything.
+  if (removed_clients_.find(client_id) != removed_clients_.end()) {
+    RAY_LOG(INFO) << "The client " << client_id << " has already been removed. This "
+                  << "should be very unusual.";
+    return;
+  }
+  removed_clients_.insert(client_id);
```

Comment:
why keep the remove_clients_, in the current design, it will keep increased.

Reply:
Some of the fault tolerance functionality that I'm working on at the moment needs to know whether a remote node is dead or alive, which is why I added this.

Reply:
I'm open to removing this (probably in a subsequent PR unless there are strong objections here), but I don't think the memory usage should be too big of a deal at the moment (since it's just 20 bytes per machine). Though I agree it'd be nice to have a better way to do this.